sharding/proposer/service.go

```diff
@@ -85,3 +123,77 @@ func (p *Proposer) proposeCollations() {
 		addHeader(p.client, collation)
 	}
 }
+
+// handleCollationBodyRequests subscribes to messages from the shardp2p
+// network and responds to a specific peer that requested the body using
+// the Send method exposed by the p2p server's API (implementing the p2p.Sender interface).
+func (p *Proposer) handleCollationBodyRequests(feed *event.Feed) {
+
+	ch := make(chan p2p.Message, 100)
+	sub := feed.Subscribe(ch)
+
+	defer sub.Unsubscribe()
+	defer close(ch)
+
+	for {
+		select {
+		case <-p.ctx.Done():
+			return
+		case req := <-ch:
+			if req.Data == nil {
+				continue
+			}
+			log.Info(fmt.Sprintf("Received p2p request from notary: %v", reflect.TypeOf(req)))
+			res, err := collationBodyResponse(req, p.client, p.shard)
+			if err != nil {
+				log.Error(fmt.Sprintf("Could not construct response: %v", err))
+				continue
+			}
+			log.Info(fmt.Sprintf("Responding to p2p request with collation with headerHash: %v", res.HeaderHash.Hex()))
+			// Reply to that specific peer only.
+			// TODO: Implement this and see the response from the other end.
+			p.p2p.Send(*res, req.Peer)
+
+		case err := <-sub.Err():
+			log.Error("Subscriber failed: %v", err)
+			return
+		}
+	}
+}
+
+// simulateNotaryRequests simulates p2p messages to the proposer that will come from
+// notaries once the system is in production. Notaries will be performing
+// this action within their own service when they are selected on a shard, period
+// pair to perform their responsibilities. This function in particular simulates
+// requests for collation bodies that will be relayed to the appropriate proposer
+// by the p2p feed layer.
+func (p *Proposer) simulateNotaryRequests(feed *event.Feed) {
```

Comment:
perhaps we should open a github issue for this (i.e. move simulateNotaryRequests to notary's `p2p.go`)

Reply:
So this was a very deliberate choice. I abstracted the meat of every object in the service into relatively small, testable unit functions that take in small interfaces as input arguments. See: `collationBodyResponse` in `p2p.go` in the `proposer` subpackage. This is done for ease of testing and because we will have end-to-end tests for greater functioning of the system.